### ISSUE TITLE
False positive 'type parameter is never used' with struct

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#60214]
Consider the following code snippet [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f59a38d253ea423eecea6ae445f618c3)

```rust
trait MyTrait<I> {}
trait SecondTrait {}

fn bar<A: MyTrait<B>, B: SecondTrait>() {}

struct MyStruct<A: MyTrait<B>, B: SecondTrait> {
    field: A
}
```
This procudes the error:

```
error[E0392]: parameter `B` is never used
 --> src/lib.rs:6:32
  |
6 | struct MyStruct<A: MyTrait<B>, B: SecondTrait> {
  |                                ^ unused type parameter
  |
  = help: consider removing `B` or using a marker such as `std::marker::PhantomData`
```

However, the parameter `B` is used - it's part of how we bound `A`.

The function and struct are equivalent in terms of generic parameters - since the equalivalent function compiles, the struct should as well.

**Comment by jonas-schievink** [comment]
This is probably intentional - the problem with type parameters that aren't used in field types is that it's not possible to infer their variance, so `PhantomData` can be used to act "as if" there's a field of a specific type incorporating the unused type parameter

**Comment by scottmcm** [comment]
Note also that, in cases like this, it means that `MyStruct { field }` can't figure out which of possibly-many `B`s you meant, so it's probably not what you want even if it were to work.

Usually this is handled by not bounding the struct, and just bounding methods on the struct.

**Comment by frewsxcv** [comment]
> the problem with type parameters that aren't used in field types is that it's not possible to infer their variance

I'm not familiar enough with compilers or programming languages to understand what it means to "infer variance", but I'm wondering if there's a more descriptive error message we can add here explaining why it's not possible. Here's an example where I hit this:

```rust
pub trait Coord {
    fn x(&self) -> f32;
    fn y(&self) -> f32;
}

pub trait Rect<C: Coord> {
    fn min(&self) -> C;
    fn max(&self) -> C;
}

pub trait Line<C: Coord> {
    fn start(&self) -> C;
    fn end(&self) -> C;
}

pub enum Geometry<C, R, L>
where C: Coord,
      R: Rect<C>,
      L: Line<C>
{
    Rect(R),
    Line(L),
}
```
